### PR TITLE
feat (api): project metadata wip

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/project_variable.go [hunk 1]
```diff
@@ -465,6 +473,11 @@ func addVariableInProjectHandler(w http.ResponseWriter, r *http.Request, db *gor
 
 	}
 
+	if err := project.UpdateLastModified(tx, c.User, p); err != nil {
+		log.Warning("updateVariablesInProjectHandler: Cannot update last modified:  %s\n", err)
+		return err
+	}
+
 	err = tx.Commit()
```
Comment (sguiheux) [inline engine/api/project_variable.go]: one line

File: engine/api/action.go [hunk 2]
```diff
@@ -153,14 +153,9 @@ func updateActionHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap,
...
-	data, errRead := ioutil.ReadAll(r.Body)
-	if errRead != nil {
-		return sdk.ErrWrongRequest
-	}
-
-	a, errJSON := sdk.NewAction("").FromJSON(data)
-	if errJSON != nil {
-		return sdk.ErrWrongRequest
+	var a sdk.Action
+	if err := UnmarshalBody(r, a); err != nil {
```
Comment (sguiheux) [inline engine/api/action.go]: pointer

File: engine/api/action.go [hunk 3]
```diff
@@ -194,15 +189,9 @@ func updateActionHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap,
...
```
Comment (sguiheux) [inline engine/api/action.go]: pointer

File: engine/api/action.go [hunk 4]
```diff
@@ -299,14 +288,8 @@ func importActionHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap,
...
 	} else { // a jsonified action is posted in body
-		btes, errRead := ioutil.ReadAll(r.Body)
-		if errRead != nil {
-			return sdk.ErrWrongRequest
-		}
-
-		a, err = sdk.NewAction("").FromJSON(btes)
-		if err != nil {
-			return sdk.ErrWrongRequest
+		if err := UnmarshalBody(r, a); err != nil {
```
Comment (sguiheux) [inline engine/api/action.go]: pointer

File: engine/api/application_group.go [hunk 5]
```diff
@@ -25,17 +23,9 @@ func updateGroupRoleOnApplicationHandler(w http.ResponseWriter, r *http.Request,
...
-	if err != nil {
-		log.Warning("updateGroupRoleOnApplicationHandler: Cannot read body :%s", err)
-		return sdk.ErrWrongRequest
-	}
-
 	var groupApplication sdk.GroupPermission
-	if err := json.Unmarshal(data, &groupApplication); err != nil {
-		log.Warning("updateGroupRoleOnApplicationHandler: Cannot unmarshal body :%s", err)
-		return sdk.ErrWrongRequest
+	if err := UnmarshalBody(r, groupApplication); err != nil {
```
Comment (sguiheux) [inline engine/api/application_group.go]: pointer

File: engine/api/application_group.go [hunk 6]
```diff
@@ -181,23 +163,14 @@ func updateGroupsInApplicationHandler(w http.ResponseWriter, r *http.Request, db
...
-	if err != nil {
-		log.Warning("addGroupInApplicationHandler> Cannot read body :%s", err)
-		return sdk.ErrWrongRequest
-	}
-
 	var groupPermission sdk.GroupPermission
-	if err := json.Unmarshal(data, &groupPermission); err != nil {
-		log.Warning("addGroupInApplicationHandler> Cannot unmarshal body :%s", err)
-		return sdk.ErrWrongRequest
+	if err := UnmarshalBody(r, groupPermission); err != nil {
```
Comment (sguiheux) [inline engine/api/application_group.go]: pointer

File: engine/api/build.go [hunk 7]
```diff
@@ -20,8 +20,8 @@ import (
 	"github.com/ovh/cds/engine/api/stats"
 	"github.com/ovh/cds/engine/api/worker"
 	"github.com/ovh/cds/engine/log"
-	"github.com/runabove/venom"
 	"github.com/ovh/cds/sdk"
+	"github.com/runabove/venom"
```
Comment (sguiheux) [inline engine/api/build.go]: wrong place

File: engine/api/group.go [hunk 8]
```diff
@@ -188,15 +179,9 @@ func getPublicGroups(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c *
...
-	// Get body
-	data, err := ioutil.ReadAll(r.Body)
-	if err != nil {
-		return sdk.ErrWrongRequest
-	}
-
 	g := &sdk.Group{}
-	if err := json.Unmarshal(data, g); err != nil {
-		return sdk.ErrWrongRequest
+	if err := UnmarshalBody(r, &g); err != nil {
```
Comment (sguiheux) [inline engine/api/group.go]: &&

File: engine/api/project.go [hunk 9]
```diff
@@ -32,15 +30,9 @@ func updateProjectHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap
...
-	// Get body
-	data, errRead := ioutil.ReadAll(r.Body)
-	if errRead != nil {
-		return sdk.ErrWrongRequest
-	}
-
 	proj := &sdk.Project{}
-	if err := json.Unmarshal(data, proj); err != nil {
-		return sdk.ErrWrongRequest
+	if err := UnmarshalBody(r, &proj); err != nil {
```
Comment (sguiheux) [inline engine/api/project.go]: &&

File: engine/api/project.go [hunk 10]
```diff
@@ -90,18 +82,10 @@ func getProjectHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c
...
-		return sdk.ErrWrongRequest
-
-	}
-
 	//Unmarshal data
 	p := &sdk.Project{}
-	if err := json.Unmarshal(data, &p); err != nil {
-		return sdk.ErrWrongRequest
-
+	if err := UnmarshalBody(r, &p); err != nil {
```
Comment (sguiheux) [inline engine/api/project.go]: &&

File: engine/api/action.go (hunk 4 above)
Comment (fsamin) [inline engine/api/action.go]: it's already pointer

**Pull request merged**